File: layers/src/main/java/com/psliusar/layers/binder/SaveWrapper.java
package com.psliusar.layers.binder;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class SaveWrapper implements Parcelable {

    public static final Creator<SaveWrapper> CREATOR = new Creator<SaveWrapper>() {
        public SaveWrapper createFromParcel(Parcel in) {
            return new SaveWrapper(in);
        }

        public SaveWrapper[] newArray(int size) {
            return new SaveWrapper[size];
        }
    };

    private final Object object;

    SaveWrapper(@NonNull Parcel in) {
        // We're recreating from serialized object. Let object be empty, because we didn't save anything.
        object = null;
    }

    public SaveWrapper(@NonNull Object obj) {
        object = obj;
    }

    public int describeContents() {
        return 0;
    }

    public void writeToParcel(Parcel out, int flags) {
        // Don't write anything. We only need our object until Parcelable gets serialized.
    }

    @Nullable
    public Object getObject() {
        return object;
    }
}

File: layers/src/main/java/com/psliusar/layers/binder/SaveWrapper.kt
package com.psliusar.layers.binder

import android.os.Parcel
import android.os.Parcelable

class SaveWrapper(
    val value: Any?
) : Parcelable {

    // We're recreating from a serialized object. Let object be empty, because we didn't save anything.
    constructor(parcel: Parcel) : this(null)

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel?, flags: Int) {
        // Don't write anything. We only need our object until Parcelable gets serialized.
    }

    companion object {
        @JvmField
        val CREATOR = object : Parcelable.Creator<SaveWrapper> {
            override fun createFromParcel(`in`: Parcel): SaveWrapper = SaveWrapper(`in`)
            override fun newArray(size: Int): Array<SaveWrapper?> = arrayOfNulls(size)
        }
    }
}